### PR TITLE
fix: Fix WebView postMessage issue

### PR DESCRIPTION
**Pull request by smilingkylan**
## **Related issues**

Fixes: https://github.com/MetaMask/mobile-planning/issues/2114

### **Before**

https://github.com/user-attachments/assets/08e86de8-f8d4-4c68-82ff-4b8f6376f117

### **After**

https://github.com/user-attachments/assets/9116a9ec-5fbb-4e35-a389-693935b008ed

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by joaoloureirop — CHANGES_REQUESTED** [review]
iOS is fixed

Android needs a fix

File: app/core/BackgroundBridge/Port.ts [hunk 1]
```diff
@@ -17,9 +18,12 @@ class Port extends EventEmitter {
...
 
-  // TODO: Replace "any" with type
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  postMessage = (msg: any, origin = '*') => {
+  postMessage = (msg: object, origin = '*') => {
+    const { activeTab, tabs } = store.getState().browser;
+    const currentActiveTab = tabs.find((tab: { id: number; url: string }) => tab.id === activeTab);
+    const currentActiveParsedUrl = new URL(currentActiveTab.url);
+    const currentActiveOrigin = currentActiveParsedUrl.origin;
+    if (currentActiveOrigin !== origin) return;
```
Comment (joaoloureirop) [inline app/core/BackgroundBridge/Port.ts]: Can we drop these changes?

**Review by Tyschenko — DISMISSED** [review]

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    sendActiveAccount = createSendActiveAccount(mockNotifyAllConnections, resolvedUrlRef);
+
+    // Setup default URL parsing mock
+    mockURLParse.mockImplementation((url: string) => ({
+      hostname: new URL(url).hostname,
+    }) as URLParse<Record<string, string>>);
+  });
+
+  describe('URL validation', () => {
+    it('should return early when no URL is provided and resolvedUrlRef is empty', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('returns early when no URL is provided and resolvedUrlRef is empty', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+      resolvedUrlRef.current = '';
+
+      await sendActiveAccount();
+
+      expect(mockURLParse).not.toHaveBeenCalled();
+      expect(mockGetPermittedEvmAddressesByHostname).not.toHaveBeenCalled();
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should use targetUrl when provided', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('use targetUrl when provided', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+      const targetUrl = 'https://example.com';
+      mockGetPermittedEvmAddressesByHostname.mockReturnValue([]);
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockURLParse).toHaveBeenCalledWith(targetUrl);
+    });
+
+    it('should use resolvedUrlRef.current when no targetUrl provided', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('use resolvedUrlRef.current when no targetUrl provided', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+      const currentUrl = 'https://current-site.com';
+      resolvedUrlRef.current = currentUrl;
+      mockGetPermittedEvmAddressesByHostname.mockReturnValue([]);
+
+      await sendActiveAccount();
+
+      expect(mockURLParse).toHaveBeenCalledWith(currentUrl);
+    });
+
+    it('should prefer targetUrl over resolvedUrlRef.current when both exist', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('use targetUrl over resolvedUrlRef.current when both exist', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockURLParse).toHaveBeenCalledWith(targetUrl);
+      expect(mockURLParse).not.toHaveBeenCalledWith(resolvedUrlRef.current);
+    });
+  });
+
+  describe('hostname extraction', () => {
+    it('should extract hostname correctly from various URL formats', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('extract hostname correctly from various URL formats', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+        );
+
+        mockURLParse.mockClear();
+        mockGetPermittedEvmAddressesByHostname.mockClear();
+      }
+    });
+  });
+
+  describe('permission checking', () => {
+    it('should check permissions for the correct hostname', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('check permissions for the correct hostname', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockGetPermittedEvmAddressesByHostname).toHaveBeenCalledWith(
+        Engine.context.PermissionController.state,
+        expectedHostname
+      );
+    });
+
+    it('should use the correct PermissionController state', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('use the correct PermissionController state', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+
+      expect(mockGetPermittedEvmAddressesByHostname).toHaveBeenCalledWith(
+        Engine.context.PermissionController.state,
+        expect.any(String)
+      );
+    });
+  });
+
+  describe('notification behavior', () => {
+    it('should not send notification when no accounts are permitted', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('don't send notification when no accounts are permitted', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+      const targetUrl = 'https://unauthorized-dapp.com';
+      mockGetPermittedEvmAddressesByHostname.mockReturnValue([]);
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should send notification when accounts are permitted', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('send notification when accounts are permitted', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockNotifyAllConnections).toHaveBeenCalledWith({
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: permittedAccounts,
+      });
+    });
+
+    it('should send notification with correct format', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('send notification with correct format', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+      await sendActiveAccount(targetUrl);
+
+      expect(mockNotifyAllConnections).toHaveBeenCalledWith({
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: permittedAccounts,
+      });
+      expect(mockNotifyAllConnections).toHaveBeenCalledTimes(1);
+    });
+
+    it('should send notification with single account', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('send notification with single account', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+
+      await sendActiveAccount(targetUrl);
+
+      expect(mockNotifyAllConnections).toHaveBeenCalledWith({
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: permittedAccounts,
+      });
+    });
+
+    it('should send notification with multiple accounts', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('send notification with multiple accounts', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+
+      expect(mockNotifyAllConnections).toHaveBeenCalledWith({
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: permittedAccounts,
+      });
+    });
+  });
+
+  describe('edge cases', () => {
+    it('should handle empty string URL gracefully', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('handle empty string URL gracefully', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+      await sendActiveAccount('');
+
+      expect(mockURLParse).not.toHaveBeenCalled();
+      expect(mockGetPermittedEvmAddressesByHostname).not.toHaveBeenCalled();
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should handle undefined URL gracefully', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('handle undefined URL gracefully', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+      await sendActiveAccount(undefined);
+
+      // Should only call if resolvedUrlRef.current has a value
+      expect(mockGetPermittedEvmAddressesByHostname).not.toHaveBeenCalled();
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should handle null URL gracefully', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('handle null URL gracefully', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      await sendActiveAccount(null as any);
+
+      expect(mockURLParse).not.toHaveBeenCalled();
+      expect(mockGetPermittedEvmAddressesByHostname).not.toHaveBeenCalled();
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should handle URL parsing errors gracefully', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('handle URL parsing errors gracefully', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+      const targetUrl = 'invalid-url';
+      mockURLParse.mockImplementation(() => {
+        throw new Error('Invalid URL');
+      });
+
+      // Should not throw, but handle gracefully
+      await expect(sendActiveAccount(targetUrl)).rejects.toThrow('Invalid URL');
+    });
+
+    it('should handle permission checking errors gracefully', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('handle permission checking errors gracefully', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+      mockGetPermittedEvmAddressesByHostname.mockImplementation(() => {
+        throw new Error('Permission check failed');
+      });
+
+      await expect(sendActiveAccount(targetUrl)).rejects.toThrow('Permission check failed');
+    });
+  });
+
+  describe('security considerations', () => {
+    it('should only send accounts for the specific hostname requested', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('only send accounts for the specific hostname requested', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+      // Test dapp2
+      await sendActiveAccount(dapp2Url);
+
+      expect(mockNotifyAllConnections).toHaveBeenCalledWith({
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: dapp2Accounts,
+      });
+    });
+
+    it('should not send accounts to unauthorized domains', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('not send accounts to unauthorized domains', async () => {
```

File: app/components/Views/BrowserTab/sendActiveAccount.test.tsx [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+      expect(mockNotifyAllConnections).toHaveBeenCalled();
+
+      mockNotifyAllConnections.mockClear();
+
+      // Unauthorized domain should not receive accounts
+      await sendActiveAccount(unauthorizedUrl);
+      expect(mockNotifyAllConnections).not.toHaveBeenCalled();
+    });
+
+    it('should prevent cross-origin account leakage', async () => {
```
Comment (joaoloureirop) [inline app/components/Views/BrowserTab/sendActiveAccount.test.tsx]:
```suggestion
    it('prevent cross-origin account leakage', async () => {
```

**Review by joaoloureirop — CHANGES_REQUESTED** [review]
code changes lgtm.

Request changes for test names to follow our contributor docs 

https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/unit-testing.md#use-it-to-specify-the-desired-behavior-for-the-code-under-test

**Review by joaoloureirop — DISMISSED** [review]

**Review by sethkfman — APPROVED** [review]
LGTM
